---
title: "Command Line Arguments"
author: "Sha Zhu"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{dEploid-Arguments}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

* `-r <R> <L>`

* `-vcf [file]` 
    File path of the isolate vcf. Assume all variants are ``PASS`` in the ``QUAL`` column, the VCF file also reqires the ``AD`` field.

** Note:
    In the current implementation, ``DEploid`` only take the first sample in the VCF file. ``DEploid`` DO NOT handle multi-allelic variants, nor indels. The ``FILTER`` column will not be used.


-plaf [*file*]
    File path of population level allele frequencies (tab-delimited plain text file), for example

.. csv-table::
    :header: CHROM, POS, PLAF
    :widths: 1,1,1

    Pf3D7_01_v3,93157,0.0190612159917058
    Pf3D7_01_v3,94422,0.135502358766423
    Pf3D7_01_v3,94459,0.156294363760064
    Pf3D7_01_v3,94487,0.143439298925837

-panel [*file*]
    File path of the reference panel (tab-delimited plain text file), for example

.. csv-table::
   :header: CHROM, POS, 3D7, Dd2, Hb3, 7G8
   :widths: 15, 10, 5, 5, 5, 5

    Pf3D7_01_v3,93157,0,0,0,1
    Pf3D7_01_v3,94422,0,0,0,1
    Pf3D7_01_v3,94459,0,0,0,1
    Pf3D7_01_v3,94487,0,0,0,1


-noPanel\
    Use population level allele frequency as prior.

.. warning::
    Flags **-panel** and **-noPanel** should not be used together.

-exclude [file]
    File path of sites to be excluded (tab-delimited plain text file).

-o [string]
    Specify the file name prefix of the output.

-k [int]
    Number of strain (default value 5).

-seed [int]
    Random seed.

-nSample [int]
    Number of MCMC samples (default value 800).

-rate [int]
    MCMC sample rate (default value 5).

-burn [float]
    MCMC

-h , -help
    Help.

-v , -version
    DEploid version.

-vcfOut\
    Save final halpotypes into a VCF file.

You may also try
----------------

-ref [file] -alt [file]
    File path of reference and alternative allele count (tab-delimited plain text file).

.. note::
    In early ``dEploid`` versions (prior to `v0.2-release`), allele counts extracted from the vcf file are placed in two files, and parsed by flags **-ref [file]** and **-alt [file]**. Tab-delimited plain text for input. First and second columns record chromosome and position labels respectively.  Third columns records the reference allele count or alternative allele count. For example,

.. csv-table:: Reference allele count
    :header: CHROM, POS, PG0390.C
    :widths: 1,1,1

    Pf3D7_01_v3,93157,85
    Pf3D7_01_v3,94422,77
    Pf3D7_01_v3,94459,90
    Pf3D7_01_v3,94487,79

.. csv-table:: Alternative allele count
    :header: CHROM, POS, PG0390.C
    :widths: 1,1,1

    Pf3D7_01_v3,93157,0
    Pf3D7_01_v3,94422,0
    Pf3D7_01_v3,94459,0
    Pf3D7_01_v3,94487,0

.. warning::
    Flags **-ref** and **-alt** should not be used with **-vcf**.

-forbidUpdateProp\
    Forbid MCMC moves to update proportions.

-forbidUpdateSingle\
    Forbid MCMC moves to update single haplotype.

-forbidUpdatePair\
    Forbid MCMC moves to update pair haplotypes.

-exportPostProb\
    Save the posterior probabilities of the final iteration of all strains.

-miss [float]
    Miss copying probability

-recomb [float]
    Constant recombination probability

-initialP [float ...]
    Initialize proportions.

-p [int]
    Output precision (default value 8).


Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
